### DISCUSSION TITLE
C# Markup + MVUX: More than one model named as <somethingModel> produces a source generator error.

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno.extensions#2307] [Q&A]
Testing C# Markup + MVUX, while trying to reproduce the Counter example, I encountered an error when adding the second partial record, the BindableViewModel is not accessible anymore. I tried this in two different projects.

```
Build started at 10:55...
1>------ Build started: Project: UnoMvux1, Configuration: Debug Any CPU ------
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>CSC : warning CS8785: Generator 'FeedsGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'UnoMvux1.MachineModel.g.cs' of the added source file must be unique within a generator.
1>c:\Test\UnoMvux1\UnoMvux1\MainPage.cs(9,30,9,47): error CS0246: The type or namespace name 'BindableMainModel' could not be found (are you missing a using directive or an assembly reference?)
1>Done building project "UnoMvux1.csproj" -- FAILED.
```

MainPage

```csharp
public sealed partial class MainPage : Page
{
    public MainPage()
    {
        this
            .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
            .DataContext(new BindableMainModel(), (page, vm) => page
                .Content(new StackPanel()
                .VerticalAlignment(VerticalAlignment.Center)
                .HorizontalAlignment(HorizontalAlignment.Center)
                .Children(
                    new StackPanel()
                        .Orientation(Orientation.Vertical)
                        .Children(
                            new TextBlock()
                                .Text("Hello Uno Platform!")
                                .FontSize(24)
                                .Margin(10),

                            new Button()
                                .Content("Click me!")
                                .Margin(10)
                        )
                )));
    }
}
```
MainModel

```csharp
internal partial record MainModel
{
    public IState<MachineModel> Machine => State.Value(this, () => new MachineModel("", ""));

    public ValueTask GetMachine() => Machine.UpdateAsync(m => m?.Load("100"));
}
```

MachineModel

```csharp
internal partial record MachineModel(string Id, string Name)
{
    public MachineModel Load(string id)
    {
        return new MachineModel(id, $"Machine {id}");
    }
}
```

[UnoMvux1.zip](https://github.com/unoplatform/uno/files/15442428/UnoMvux1.zip)

**Comment by mcNets** [comment]
If I rename the machine model as `internal partial record Machine(string Id, string Name)` the error dissapear.

Having a look at binlog I found this warning (text could be innacurated, I translated it):

```
System.ArgumentException: The hintName item "OsmoUI.Models.MaquinaModel.g.cs" of the added source file must be unique within a generator. (Parameter 'hintName')
at Uno.Extensions.Reactive.Generator.FeedsGenerator.Execute(GeneratorExecutionContext context)
at Microsoft.CodeAnalysis.SourceGeneratorAdaptor.<Initialize>b__5_5(SourceProductionContext productionContext, GeneratorContextBuilder contextBuilder)
at Microsoft.CodeAnalysis.UserFunctionExtensions.<>c__DisplayClass3_0`2.<WrapUserAction>b__0(TInput1 input1, TInput2 input2, CancellationToken token)

```

**Reply by dr1rrb** [reply]
Humm on my side I'm not even able to compile the project until I rename the `MachineModel` to something else. The `BindableMainModel` is not being generated otherwise 🤔 (The generator probably crashing for the same reason as you).

So this confirms my hypothesis : we have a name conflict between your `McahineModel` and stuff generated for the `public IState<MachineModel> Machine` in the `MainModel`. And actually changing the `MachineModel` to be `class` instead of a `record` works ...

Will investigate it.

**Reply by dr1rrb** [reply]
Ok so:

1. As the type end by `Model` we are generating the `BindableMachineModel` (to be used as `DataContext` of a `Page`);
2. As the `MachineModel` is being used in a `State<MachineModel>` (in `MainModel`) **and** it is a `record`, we are generating a `BindableMachineModel` to support 2-way bindings on sub-properties. This one being designed to be used as a property of the `BindableMainModel`.

The issue is that both are being generated with file name `MachineModel.g.cs` AND they are not compatible!

#  WORKAROUND
As in this example we don't need the first kind of generation, on the `MachineModel` add the attribute `[Uno.Extensions.Reactive.ReactiveBindable(false)]`

> Note: Changing the `MachineModel` to be a `class` instead of a `record` will disable second kind of generation, but it's not expected here.

**Reply by dr1rrb** [reply]
Created an issue to track that https://github.com/unoplatform/uno.extensions/issues/2310

**Comment by sasakrsmanovic** [comment]
@agneszitte  Do you think we can take a look at this please?

**Reply by dansiegel** [reply]
cc: @dr1rrb
